[PallonCX] iP

PallonCX opened the pull request:
# Killua

> "Just Let Me Handle This." - Killua Zoldyck ([source](https://www.cbr.com/hxh-best-killua-quotes/#quot-if-you-so-much-as-touch-them-i-39-ll-kill-you-quot))

Killua frees your mind of having to remember things you need to do. It's,
- Text-based
- Easy to learn
- FAST SUPER FAST to use

There are three type of tasks Killua can manage for now:
1. **Todo** task - Support task description.
2. **Deadline** task - Support task description, due date, and due time (time is optional).
3. **Event** task - Support task description, start date and time, end date and time (time is optional).

Killua can help you to store the tasks in your local device's memory. The only thing you need to do is to modify the file path in the code below (the code can be found in `Duke.java`):
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

## Future RoadMap :sunglasses:

Features below may be implemented soon:
- [ ] Better Graphical User Interface (GUI).
- [ ] Higher tolerance to command given, and more acceptable commands.
- [ ] Applying Natural Language Processing (NLP) to improve user interaction experience.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Duke.java. Add Storage.java, Ui.java

PallonCX committed:
Add Task.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Add Deadline.java

PallonCX committed:
Add Event.java

PallonCX committed:
Add Todo.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Task.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify EXPECTED.TXT

PallonCX committed:
Modify input.txt

PallonCX committed:
Add DukeException.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify Todo.java

PallonCX committed:
Modify Task.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify Event.java

PallonCX committed:
Modify Deadline.java

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify DukeException.java

PallonCX committed:
Modify TaskManager.java

PallonCX committed:
Add data

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify Deadline.java

PallonCX committed:
Modify DukeException.java

PallonCX committed:
Modify Event.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify TaskManager.java

PallonCX committed:
Modify duke.txt

PallonCX committed:
Modify AddCommand.java

PallonCX committed:
Modify Command.java

PallonCX committed:
Modify DeleteCommand.java

PallonCX committed:
Modify EditCommand.java

PallonCX committed:
Modify ExitCommand.java

PallonCX committed:
Modify ListCommand.java

PallonCX committed:
Add Parser.java

PallonCX committed:
Add data

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify DukeException.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify TaskList.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Refactor files into packages

PallonCX committed:
Merge branch 'add-gradle-support'

PallonCX committed:
Add src/test

PallonCX committed:
Modify build.gradle

PallonCX committed:
Add JavaDoc to all files

PallonCX committed:
Modify Parser.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify AddCommand.java

PallonCX committed:
Add FindCommand.java

PallonCX committed:
Modify Parser.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify data/tasks.txt

PallonCX committed:
Merge branch 'branch-A-CodingStandard'

PallonCX committed:
Modify FindCommand.java

PallonCX committed:
Modify ListCommand.java

PallonCX committed:
Modify Ui.java

tankh99 commented on data/tasks.txt:
Try to avoid committing storage files

PallonCX replied:
If you are referring to the file that stores the tasks, actually I'm also considering to add them to gitignore. As far as I'm concerned, we are not given explicit instructions on what should be added to gitignore, please tell me if you know where to get more information about this.

tankh99 replied:
Apologies, this wasn't part of the course. I was just giving advice in general

PallonCX replied:
No bro please don't apologise, I feel so bad for it hahaha. I'm really curious to know what should be gitignore while I'm doing the works. Since you mention that it wasn't part of the course, perhaps I will ask at forum to get some general advices from the teaching team. This point that you have brought up is absolutely great and useful, thanks a lot!

tankh99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,220 @@
+        sb.append(home);
+
+        // Ensure that the file exist
+        for (int i = 0; i < splitedFile.length; i++) {
+            sb.append("/" + splitedFile[i]);
+            Path path = Paths.get(sb.toString());
+
+            // Folder not exist
+            if (i != splitedFile.length - 1 && !Files.exists(path)) {
+                try {
+                    Files.createDirectory(path);
+                } catch (Exception e) {
+                    throw new DukeException(e.getMessage());
+                }
+            }
+
+            // Folder exist but not folder
+            if (i != splitedFile.length - 1 && !Files.isDirectory(path)) {
+                throw new DukeException();
+            }
+
+            // File not exist
+            if (i == splitedFile.length - 1 && !Files.exists(path)) {
+                try {
+                    Files.createFile(path);
```
Why did you create a for loop just to check if each directory exists? I think that System.getProperty("user.dir") already ensures that the path it returns all exists

PallonCX replied:
My concern is regarding the file path given by user, e.g. for a file path `folderA/folderB/fileC`, indeed System.getProperty("user.dir") already ensures that the path it returns all exists, but when I append the file path given by the user, it could happen certain errors such as:

1. folderA and / or folderB does not exist or exist as a file with same name
2. fileC does not exist or exist as a folder with same name

Therefore, I decide to handle this potential problem by performing checks on each layer, and create a file at the given file path iff there is no error as mentioned above.

tankh99 replied:
You could just append the path to user.dir and then check use File.exists() method once to see it's valid. A for loop is most likely unnecessary here

tankh99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,220 @@
+                        String deadline = splited[3];
+                        String[] splitedDateTime = deadline.split(" ");
+
+                        // Date and time
```
nit: Perhaps give a more descriptive comment

tankh99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,220 @@
+
+        try {
+            FileWriter fw = new FileWriter(filePath.toString());
+            fw.write(textToAdd.toString());
+            fw.close();
```
I think you should close fw in a finally block. Because this would not close in case an exception is thrown

PallonCX replied:
That is a good catch, will fix it later, thanks!

tankh99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,348 @@
+ */
+public class Parser {
+
+    public enum TaskType {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND, UNKNOWN
+    }
```
I think it could be good if this was extracted out to a class of its own

PallonCX replied:
I'm not sure whether it is a common practise to put enum as a class of its own or make it part of some other class, but sounds like a valid suggestion as enums are kind of similar to classes, thanks!

tankh99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,348 @@
+     * @return A command object to be executed.
+     * @throws IllegalArgumentException If any syntax error exists.
+     */
+    public static Command handleTaskBye(String[] splitedTask) throws DukeException {
+        // Incorrect command syntax handler
+        if (splitedTask.length > 1) {
+            throw new DukeException("Syntax of bye: bye");
+        }
+        return new ExitCommand();
+    }
```
Although you're following naming convention, perhaps you could make the function more intuitive. E.g. handleTaskExit since "bye" could potentially be replaced with any other word, but the action would remain the same "exit application"

tankh99 commented on src/main/java/duke/command/Command.java:
I like abstraction and I like how you implemented it in this class

tankh99 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+
+    /**
+     * Returns True if and only if the command is `ExitCommand`.
+
```
nit: missing a *

PallonCX replied:
Ah yes

tankh99 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,59 @@
+        storage.saveChanges(tasks);
+    }
+
+    /**
+     * Returns False.
+
+     * @return False.
+     */
```
Could use a more descriptive description. E.g., what does isExit mean and what is it used for? It's not immediately apparent

nit: also missing another *

PallonCX replied:
Ooh yea thanks!

tankh99 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return The formatted string containing done status and description of the task.
+     */
+    @Override
+    public String getDescriptionStatus() {
+        return "[D]" + (isDone ? "[X] " : "[ ] ") + this.description + " (by: "
+                + this.getDeadlineForDisplay() + ")";
+    }
```
If you're overriding, you could actually create a toString() method in Task and then use that toString() method and modify it inside Deadline.

PallonCX replied:
I choose to create a new method named `getDescriptionStatus` instead of using the original `toString` inherit from Object class because I'm not sure how many customised strings I will need to make for the class along the way in future increments.

tankh99 commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class DeadlineTest {
+    @Test
+    public void getDeadlineForDisplayTest(){
```
Does not follow standard test naming convention. Correct me if I'm wrong if this isn't the standard convention

`featureName_situation_expectedResult`. 

E.g. `getDeadline_validFormat_returnsFormattedDate`

PallonCX replied:
Oh yea you are right, I'm so frustrated while setting up the Gradle and making the JUnit tests, until a point that I only want it to work, which makes me forget the naming convention haha.

tankh99 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,172 @@
+        storage.saveChanges(tasks);
+    }
+
+    /**
+     * Returns False.
+
+     * @return False.
+     */
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
if the superclass method already has a javadoc comment, no need to write comment for overridden method

PallonCX replied:
Oww alright, I saw from somewhere (can't recall) that we can use some tag, something like @inherit..., but I didn't manage to find it in my IDE. Any clue about it?

tankh99 replied:
I dont think such an annotation exists? You just don't have to write any comment at all for overridden methods

tankh99 reviewed (commented):
I've reviewed your PR briefly, found just a bunch of nits. Mainly confused about your storage code and why you decided to do the things you did there

I also found a few coding standard violations but not a whole lot. You added comments to almost if not every function and class, good job! Although, I'm of the opinion that some comments can be left out if they're not particularly useful. And also, some comments that you wrote were not particularly descriptive either, like "isExit"

simkianseng commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * The `DukeException` class represents any exception related to the Duke program.
+ */
+public class DukeException extends Exception{
```
Perhaps this was missed, the `Exception{` could have a space between the word `Exception` and the brace `{`.

> White space within a statement &ndash; se-education.org

Can refer to the documentation [here](https://se-education.org/guides/conventions/java/index.html).

PallonCX replied:
Ooh yea you are right!

simkianseng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,348 @@
+        task = task.substring(9);
+
+        // Split the string with /by
+        String[] splitedBy = task.split(" /by ");
```
Seems like the variable `splitedBy` was misspelled, perhaps can refer to W4.6c of the notes [here](https://nus-cs2103-ay2324s2.github.io/website/book/codeQuality/nameWell/basic/useStandardWords/). This was found in other areas of the codebase as well.

PallonCX replied:
Hmmm my original meaning for the variable is "splited string by the keyword '/by'". What do you think a better name will be?

simkianseng replied:
I think the past tense of "split" is "split". I don't think there is such a word as "splited". Hence the variable name can be called `splitBy`, but no worries just a minor issue.

simkianseng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,220 @@
+        Path path = Paths.get(home, this.filePath);
+
+        // Read from file
+        try {
```
It seems that your code have deep nesting. Perhaps the code block can be simplified? Here is an example of a deeply nested (or arrowhead style) code:
```java
int subsidy() {
    int subsidy;
    if (!age) {
        if (!sub) {
            if (!notFullTime) {
                subsidy = 500;
            } else {
                subsidy = 250;
            }
        } else {
            subsidy = 250;
        }
    } else {
        subsidy = -1;
    }
    return subsidy;
```
Refer [here](https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting) for related documentation.

PallonCX replied:
Hmm it's quite hard for me to simplify it due to the design of my code. To be more precise:
1. The `try` block is required to catch exceptions.
2. The `while` block is required to read inputs continuously.
3. The `for` block is required to consider every task.
4. The `switch` block is required to decide which command to perform.
5. Some `try` block deep inside is indeed redundant as there is a big `try` block at very outside.
6. Some `if-else` block is required to check for specific condition.

Feel free to tell me if you have any specific suggestion, thanks!

simkianseng replied:
1. Perhaps combine multiple `try...catch` blocks into one? I also noticed that you catch `Exception` instead of specifying the specific exception to catch (e.g. `IOException` or `FileNotFoundException`) which is something we what to avoid (I subconsciously remembered in one of the briefing video @damithc mentioned something along the lines of "Don't catch them all" but I can't seem to find it in the textbook). Also, take note to catch all checked exceptions (refer to W1.5 [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week1/topics.html#W1-5)) The following example might also fix your issues.
```java
try {
    <your code>
} catch (IOException e) {
    <your code>
} catch (FileNotFoundException e) {
    <your code>
}
```
2./3./4./5./6. Perhaps the blocks can be abstracted into smaller methods.

Not really sure if mine is a good example but maybe you can take some inspiration [here](https://github.com/SimKianSeng/ip/blob/master/src/main/java/simpli/storage/Storage.java)?

As everyone has different implementations it's quite hard for me to tell you what is right, but I can point out what doesn't seem right to me and maybe provide you with the reference materials. Hope this was helpful.

PallonCX replied:
Thanks for your nice elaboration, I will try to apply them in my code!

simkianseng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,220 @@
+                                Task newTask = new Event(content, isDone, parsedFromDateTime, parsedToDateTime);
+                                items.add(newTask);
+                            } else if (splitedFrom.length == 2 && splitedTo.length == 1) {
+                                DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Perhaps can the variable name `formatter1` and `formatter2` be something more meaningful? Maybe something like `formatterToDate` and `formatterFromDate` respectively? Refer [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/topics.html#implementation-code-quality-naming-intermediate-use-name-to-explain) for more documentation.

PallonCX replied:
Sounds like a valid one, I feel too tedious when setting up these part hahaha. Thanks!

simkianseng commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,239 @@
+     */
+    @Override
+    public String[] getFields() {
+        String[] result = new String[4];
```
Perhaps the variable `result` can be in plural?
> Plural form should be used on names representing a collection of objects. &ndash; se-education.org

Here is an example:
```java
Collection<Point> points;
int[] values;
```
Refer for the documentation [here](https://se-education.org/guides/conventions/java/intermediate.html).

PallonCX replied:
Ah true!

simkianseng reviewed (commented):
Minor code quality issues. Overall good codebase. Good job! 😄

PallonCX committed:
Modify Parser.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify AddCommand.java

PallonCX committed:
Add TaskType.java

PallonCX committed:
Modify DukeException.java

PallonCX committed:
Modify Storage.java

PallonCX committed:
Modify Command.java

PallonCX committed:
Modify DeleteCommand.java

PallonCX committed:
Modify Event.java

PallonCX committed:
Modify DeadlineTest.java

PallonCX committed:
Modify EventTest.java

PallonCX committed:
Fix code style

PallonCX committed:
Add DialogBox.java

PallonCX committed:
Add Launcher.java

PallonCX committed:
Add Main.java

PallonCX committed:
Add MainWindow.java

PallonCX committed:
Add resources

PallonCX committed:
Modify Duke.java

PallonCX committed:
Modify Ui.java

PallonCX committed:
Modify AddCommand.java

PallonCX committed:
Modify Command.java

PallonCX committed:
Modify DeleteCommand.java

PallonCX committed:
Modify EditCommand.java

PallonCX committed:
Modify ExitCommand.java

PallonCX committed:
Modify FindCommand.java

PallonCX committed:
Modify ListCommand.java

PallonCX committed:
src/main/java/duke/task/Task.java: add assert

The field isDone is only possible to be 'Y' which is yes or 'N'
which is no.

Java assert is used to ensure that it can only be 'Y' or 'N'.

Java assert is used instead of throwing exception because the
previous line assigns 'Y' or 'N' to isDone, and hence if isDone
is not 'Y' or 'N' at this point, that must be some mistakes
made by programmer, which is suitable to be handled with
assertion.

PallonCX committed:
Code quality: Refactor

Code quality is important to improve readability.

Codes are refactored such as magic literals are placed as constant,
and some identical code snippets in alternative branches are combined.

This change can simplify the code, to ease reading and understanding,
and the removal of magic literals reduce potential bugs.

PallonCX committed:
Ui.java: Use Stream for find function

By using Stream to process the items, not only we can use
convenient functionality such as filter, we can also use different
stream operations on the items.

Let's use Stream instead of for-loop.

Stream can replace the usage of for-loop.

PallonCX committed:
Merge pull request #4 from PallonCX/branch-A-Assertions

src/main/java/duke/task/Task.java: add assert

PallonCX committed:
Merge branch 'master' into branch-A-CodeQuality

PallonCX committed:
Merge pull request #2 from PallonCX/branch-A-CodeQuality

Code quality: Refactor

PallonCX committed:
Merge branch 'master' into branch-A-Streams

PallonCX committed:
Merge pull request #3 from PallonCX/branch-A-Streams

Ui.java: Use Stream for find function

PallonCX committed:
Add .github folder

PallonCX committed:
Add TagCommand

PallonCX committed:
Highlight user input error

PallonCX committed:
Add background for text

PallonCX committed:
Add comment

PallonCX committed:
Add title bar

PallonCX committed:
Add Ui.png

PallonCX committed:
Create user guide

PallonCX committed:
Fix code style

PallonCX committed:
Add tests for Deadline class

PallonCX committed:
Add more tests
